csrc/C++20/ranges

```diff
@@ -0,0 +1,78 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <type_traits>
+
+// This is backport of C++20 ranges library to C++17.
+
+namespace std {
```

Comment:
I think it's best to not add this to `std`. `nvfuser::views::iota_views` is better. `std::views` gives readers the wrong impression that this is defined in standard headers. It's not and it's not even a copy of it as you mentioned.

Reply:
I would prefer to use `std`, because the purpose of this is to provide a smooth upgrade to C++20. I want to be able to just remove the `ranges` header and the code will just work. Otherwise, there is no point for backporting C++20 while maintaining the same interface.

Reply:
Defining a class that is not defined in standard library in `std` namespace is supported use case.

Reply:
https://en.cppreference.com/w/cpp/language/extending_std

Reply:
We have been doing this same thing:
https://github.com/NVIDIA/Fuser/blob/main/csrc/C%2B%2B20/compare
https://github.com/NVIDIA/Fuser/blob/main/lib/dynamic_type/src/dynamic_type/C%2B%2B20/type_traits

Reply:
>  It's not and it's not even a copy of it as you mentioned.
It is not possible to provide a 100% copy because of missing language features in C++17. But here, all the used cases should have exactly the same interface, just like if you are using C++20.

Reply:
> Defining a class that is not defined in standard library in std namespace is supported use case.

Yes, I totally understood it's supported. My concern is about giving readers the wrong impression that this is the real `std::views` while it's not. For example, I wouldn't even suspect a potential bug in our implementation. `nvfuser::views` makes that explicit and clear. 

> we had a c10::optional vs std::optional for the C++14 upgrade, and our codebase was mixing both of them

I'm surprised it was a problem for us. I can run `git grep -l 'c10::optional' | xargs sed -i 's/c10::optional/std::optional/g'` and send out a PR in five minutes. 

In fact, `nvfuser::views` helps you upgrade to C++20. Because it works with both C++17 and C++20, and we only bother replacing it after we decide to require C++20.

Reply:
> I'm surprised it was a problem for us.
But PyTorch is using `c10::optional`, so all the interfaces with pytorch should use `c10::optional`, and just doing a naive search and replace does not work.

The overall thing I want to do is to close the gap between C++17 and C++20 in the standard library. I want to preserve the capability of just swapping the `-std=c++17` with `-std=c++20` to quickly test forward compatibility. If I am not allowed to put things into `namespace std`, I would prefer to just close this PR because there are multiple ways to implement `irange` on C++17, and the implementation of `irange` in this PR is not even a good one without being in `namespace std`, compared to other ways of implementing `irange`.

Reply:
This is a cleanup PR, and just like any other cleanup PRs, it is meant to improve developer experience. I have a feeling that building up the `namespace std` is an improvement of developer experience (at least for me). Building up the `namespace std` is my main goal, and I have multiple minor applications I want to use it with. Cleaning up `irange` is only a way to test the build-up. If people don't think this is an improve, then I will close this PR. To me this is not my main task, it is just something that "do one work today, and in the future, I will feel happier". If this thing is not in `namespace std`, I will be as unhappy with it as I am with `c10::irange`, because I am not interested in building our own ranges library.